packages/tpetra/core/src/Tpetra_idot.hpp

```diff
@@ -287,6 +299,14 @@ lclDotRaw (typename ::Tpetra::MultiVector<SC, LO, GO, NT>::dot_type* const resul
         auto Y_lcl_1d = subview (Y_lcl, rowRange, 0);
         auto result_0d = subview (result, 0);
         KokkosBlas::dot (result_0d, X_lcl_1d, Y_lcl_1d);
+
+        // FENCE REVIEW - CONFIRMED FAILURE
+        //   Testing: This code is exercised by unit tests.
+        //   GTX960:  Failed with CUDA_LAUNCH_BLOCKING=0 and fence removed.
+        //   White:   Not checked since failure already confirmed on GTX960.
+        //   Plan:    May consider downstream failure point and move fence.
+        //   Notes:
+
```

Comment:
I've confirmed that if this fence and the one below are commented out, and CUDA_LAUNCH_BLOCKING=0, these tests fail on vortex:
```
	 39 - TpetraCore_idot_MPI_4 (Failed)
	 44 - TpetraCore_Core_initialize_where_user_initializes_mpi_MPI_4 (Failed)
	 45 - TpetraCore_Core_ScopeGuard_where_user_initializes_mpi_MPI_4 (Failed)
```

Reply:
I think that this is a problem from KokkosBlast::dot. I believe that the dot uses parallel reduce and it should wait until the result is fully updated even though it uses a device result view.

Reply:
Just a reminder we had this discussion regarding dot:
[https://github.com/kokkos/kokkos-kernels/pull/660](https://github.com/kokkos/kokkos-kernels/pull/660)

Reply:
Note that only the TpetraCore_idot_MPI_4 failure is specifically related to this fence.

Reply:
Oh...... I have a short-term memory and already said the opposite in that issue.

Reply:
Based on @brian-kelley's suggestion in today's Tpetra meeting, I put fences just before the calls to `ireduce` in `Tpetra_idot.hpp`.  (Brian said that there needed to be a sync before calling MPI.)   With these fences, the `idot` test passes.

Reply:
@jhux2 Thanks. Let me know if you want me to PR that change.

Reply:
@jhux2 OK, good that it passes. I found this old discussion on KokkosKernels about whether KokkosBlas::dot should fence after the case of reducing to a 1-D view: https://github.com/kokkos/kokkos-kernels/pull/660 . The concensus there was that the caller of KokkosBlas::dot should fence if they're going to read the UVM result on host.

I think that if we could replace the fence you just added with a deep_copy instead, that makes the intent of the code clearer. You can deep_copy from a UVM view to its host mirror (itself) and Kokkos will just fence the execution space once. It would be identical performance but makes the intent clearer than a fence.

Something else is a little funky in the lclDotRaw():
```
typedef View<dot_type*, device_type> result_dev_view_type;
typedef typename result_dev_view_type::HostMirror result_host_view_type;

// later:
  result_dev_view_type result;
  result_host_view_type result_h;
  if (resultOnDevice) {
    result = result_dev_view_type (resultRaw, numVecs);
    if (runOnHost) {
      // create_mirror_view doesn't promise to create a deep copy, so we
      // can't rely on this case to avoid an extra buffer if the input
      // communicator is an intercomm.
      result_h = ::Kokkos::create_mirror_view (result);
    }   
    // If running on device, not host, then we don't use result_h.
  }
  else {
    result_h = result_host_view_type (resultRaw, numVecs);
    if (! runOnHost) { // running on device, not host
      // We have to do a separate allocation here, since resultRaw is
      // not accessible from device.
      result = result_dev_view_type ("result", numVecs);
    }   
    // If running on host, not device, then we don't use result.
  }
```
BTW, I'm pretty sure we should have ``Kokkos::MemoryTraits<Kokkos::Unmanaged>`` in the typedef result_dev_view_type, and use ``View<dot_type*, HostSpace, MemoryTraits<Unmanaged>`` for ``result_host_view_type``. We're just trying to write the result to a C-style array that either lives on device or host. I asked Christian T about what it's doing now on kokkos slack.

Reply:
After looking some more, I think there's no avoiding the fence before ireduce/iallreduce. The fences inside lclDotRaw can go.

Reply:
@brian-kelley So do you think a fence is better than a deep_copy?

Reply:
@jhux2 My thought was that since there's only one view that needs to be synchronized for access on host, a deep_copy is nicer than just a fence because a fence in the code doesn't tell you why it's there. A deep_copy from UVM to itself is internally just a fence, but the intent of accessing the data on host is clear without needing comments.

Reply:
But now, I'm not sure that's a good idea because if we compute on device and communicate from device, but have CUDA_LAUNCH_BLOCKING off, then you still need to fence before the communication, but the current code won't go into the if for if communication is needed. So I think just forget about this and keep what you had (one fence before the communication)

Reply:
Is it really necessary to write lclDotRaw with this much of complexity ? With several deep copy (mostly mirroring), removing or adding one fence does not seem to affect overall performance except for a lucky case that can avoid all deep copies in the code.

Reply:
@kyungjoo-kim Yes, I'm overthinking it :)

Let's just fence the execution space before the communication and call it done.

Later if we want we can simplify things a bit: lclDotRaw might as well always put the result on host, there's no performance benefit at all to the "resultOnDevice" stuff. If it's computed on device, it needs to get copied to host either by deep copy or UVM.

Reply:
@brian-kelley Thank you for the explanation.  @MicheldeMessieres Let's put fences before each ireduce/iallreduce.